sdks/go/pkg/beam/util/starcgenx/starcgenx.go

```diff
@@ -0,0 +1,565 @@
+	imports map[string]struct{}
+
+	allExported bool // Marks if all ptransforms are exported and available in main.
+}
+
+// Summary prints out a summary of the shims and registrations to
+// be generated to the buffer.
+func (e *Extractor) Summary() {
+	e.Print("\n")
+	e.Print("Summary\n")
+	e.Printf("All exported?: %v\n", e.allExported)
+	e.Printf("%d\t Functions\n", len(e.functions))
+	e.Printf("%d\t Types\n", len(e.types))
+	e.Printf("%d\t Shims\n", len(e.funcs))
+	e.Printf("%d\t Emits\n", len(e.emits))
+	e.Printf("%d\t Inputs\n", len(e.iters))
+}
+
+// lifecycleMethodName returns if the passed in string is one of the lifecycle method names used
+// by the Go SDK as DoFn or CombineFn lifecycle methods. These are the only methods that need
+// shims generated for them, as per beam/core/graph/fn.go
+// TODO(lostluck): Move this to beam/core/graph/fn.go, so it can stay up to date.
+func lifecycleMethodName(n string) bool {
+	switch n {
+	case "ProcessElement", "StartBundle", "FinishBundle", "Setup", "Teardown", "CreateAccumulator", "AddInput", "MergeAccumulators", "ExtractOutput", "Compact":
+		return true
+	default:
+		return false
+	}
+}
+
+// Bytes forwards to fmt.Fprint to the extractor buffer.
+func (e *Extractor) Bytes() []byte {
+	return e.w.Bytes()
+}
+
+// Print forwards to fmt.Fprint to the extractor buffer.
+func (e *Extractor) Print(s string) {
+	if e.debug {
+		fmt.Fprint(&e.w, s)
+	}
+}
+
+// Printf forwards to fmt.Printf to the extractor buffer.
+func (e *Extractor) Printf(f string, args ...interface{}) {
+	if e.debug {
+		fmt.Fprintf(&e.w, f, args...)
+	}
+}
+
+// FromAsts analyses the contents of a package
+func (e *Extractor) FromAsts(imp types.Importer, fset *token.FileSet, files []*ast.File) error {
+	conf := types.Config{
+		Importer:                 imp,
+		IgnoreFuncBodies:         true,
+		DisableUnusedImportCheck: true,
+	}
+	info := &types.Info{
+		Defs: make(map[*ast.Ident]types.Object),
+	}
+	if len(e.Ids) != 0 {
+		// TODO(lostluck): This becomes unnnecessary iff we can figure out
+		// which ParDos are being passed to beam.ParDo or beam.Combine.
+		// If there are ids, we need to also look at function bodies, and uses.
+		var checkFuncBodies bool
+		for _, v := range e.Ids {
+			if strings.Contains(v, ".") {
+				checkFuncBodies = true
+				break
+			}
+		}
+		conf.IgnoreFuncBodies = !checkFuncBodies
+		info.Uses = make(map[*ast.Ident]types.Object)
+	}
+
+	if _, err := conf.Check(e.Package, fset, files, info); err != nil {
+		return fmt.Errorf("failed to type check package %s : %v", e.Package, err)
+	}
+
+	e.Print("/*\n")
+	var idsRequired, idsFound map[string]bool
+	if len(e.Ids) > 0 {
+		e.Printf("Filtering by %d identifiers: %q\n", len(e.Ids), strings.Join(e.Ids, ", "))
+		idsRequired = make(map[string]bool)
+		idsFound = make(map[string]bool)
+		for _, id := range e.Ids {
+			idsRequired[id] = true
+		}
+	}
+	// TODO(rebo): Need to sort out struct types and their methods, so we only
```

Comment:
rebo -> lostluck (in the file)

Reply:
Done. And this comment was addressed, so it could be removed.